specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -35,369 +35,401 @@ https://www.dolthub.com/blog/2020-05-13-dolt-commit-graph-and-structural-sharing
-  - Increment `cursor.parent.index` by `1`
-  - Set the `cursor.parent.links` at `cursor.parent.index` to the new `CID`
-  - Set the `cursor.parent.keys` at `cursor.parent.index` to node.keys[0]
-  - Set cursor.node to `node`
-  - Set cursor.index to `0`
-- Return `RebalanceTree(cursor.parent)`
-
-### Put(ProllyNode, key, value) : ProllyNode
-
-- Get `cursor` from `CursorAtItem(ProllyNode, key)`
+  - If `CursorIsAtEnd(cursor)`
+    - If `cursor.parent` is `null` or `ShouldCreateBoundry(ProllyTreeConfig, cursor)` is `true`
+      - break the loop
+    - Else, if `cursor.parent` has another prolly node
+      - merge it with `cursor.node`
+    	- break the loop
+  - If `ShouldCreateBoundry(ProllyTreeConfig, cursor)` is `false`
+  	- Call `AdvanceCursor(cursor)`
+  	- continue to next loop cycle
+	- Call `SplitCursor(cursor)`
+- Return `RebalanceTree(cursor.parent, ProllyTreeConfig)`
+
+### Put(ProllyTree tree, key, value) : ProllyTree
+
+This is a public facing API for setting keys in the prolly tree.
+
+- Get the `config` from the `tree` using `load(tree.config)`
+- Get root `node` using `load(tree.root)`
+- Get `cursor` from `CursorAtItem(node, key)`
 - If `CursorIsValid(cursor)` is `true`
 	- check if the key is at `CursorGetKey`
 	- If it is
 		- Set the value in the `cursor.node.values` at `cursor.index` to `value`
 	- If it isn't
 		- add the `key` to `cursor.node.keys` after the `cursor.index`, shifting subsequent items down
 		- add the `values` to `cursor.node.values` after the `cursor.index`, shifting subsequent items down
+		- increment `cursor.index` by `1`
+	- Get the `length` from `save(cursor.node).bytes`
+	  - If `length` > `config.maxChunkSize`
+	  - call `SplitCursor(cursor)`
 - TODO: If false, how is this reached? It means that there were no keys even "close" to `key`?
-- return `RebalanceTree(cursor)`
+- get a new `root` from  `RebalanceTree(cursor, config)`
+- Create a new `updatedTree` by duplicating `tree`
+- Set `updatedTree.root` to `root`
+- Return `updatedTree`
+
+### Delete(ProllyTree tree, key) : ProllyTree
 
-### Delete(ProllyNode, key, value) : ProllyNode
+Removes a key from a ProllyTree if it exists
 
-- Get `cursor` from `CursorAtItem(ProllyNode, key)`
+- Get the `config` from the `tree` using `load(tree.config)`
+- Get root `node` using `load(tree.root)`
+- Get `cursor` from `CursorAtItem(TreeNode, key)`
 - If `CursorIsValid(cursor)` is `false`
-	- Return the `ProllyNode`
+	- Return the `TreeNode`
 - check if the key is at `CursorGetKey(cursor)`
 - If it is
 	- Remove the key in `cursor.node.keys` at `cursor.index`
 	- Remove the value in `cursor.node.values` at `cursor.index`
-- return `RebalanceTree(cursor)`
+- get a new `root` from  `RebalanceTree(cursor, config)`
+- Create a new `updatedTree` by duplicating `tree`
+- Set `updatedTree.root` to `save(root).cid`
+- Return `updatedTree`
 
-### Search(ProllyNode, prefix) : Iterator<key, value>
+### Search(TreeNode, prefix) : Iterator<key, value>
```

Comment:
There is still the problem that you cannot iterate over arbitrary ranges. The problem is the step `Close the iterator and return`. It will automatically stop when the prefix doesn't match, so it's not fully up to the application.

I propose changing `prefix` to `start` and don't automatically stop somewhere, but just keep iterating to the end, unless the application stops the iteration.

Reply:
Sounds good 👍

I've mostly been using prefix based search with stuff like leveldb, but I can see the point in focusing on start instead of prefix.